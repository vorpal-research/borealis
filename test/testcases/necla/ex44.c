#include <stdlib.h>

#define INIT_SIZE 100

typedef struct _vector
{
	size_t size;
	size_t capacity;
	int *data;
} Vector;

Vector* create_vector()
{
	Vector *node;

	node = (Vector *) malloc(1 * sizeof(Vector));
	node->size = 0;
	node->capacity = INIT_SIZE;
	node->data = (int *)malloc(node->capacity* sizeof(int));
        ASSUME(node->data);
	return node;
}

void push_back(Vector *vct, int value)
{
	int *tmp;
        ASSERT(vct -> size <= vct -> capacity);
	if (vct->size >= vct->capacity)
	{
		vct->capacity += INIT_SIZE;
		tmp = (int*) malloc (vct->capacity * sizeof(int));
                ASSUME(tmp != NULL);
		memcpy(tmp, vct->data, (vct->capacity - INIT_SIZE)*sizeof(int));
		free(vct->data);
		vct->data = tmp;
	}
        
	vct->data[vct->size++] = value;
}

int main(int n){
   int i;
   Vector * a;
   ASSUME(n > 0);
   
   a = creat_vector();
   
   for (i =0; i<n; i++) 
   {
       push_back(a, i);
   }
   
   ASSERT(a->size == n);

   return 1;
}
