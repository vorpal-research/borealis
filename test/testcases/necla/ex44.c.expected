
[
	
	{
		"location" : 
		{
			"filename" : "test/testcases/necla/ex44.c",
			"loc" : 
			{
				"col" : 5,
				"line" : 30
			}
		},
		"type" : "BUF-01"
	},
	
	{
		"location" : 
		{
			"filename" : "test/testcases/necla/ex44.c",
			"loc" : 
			{
				"col" : 5,
				"line" : 32
			}
		},
		"type" : "BUF-01"
	},
	
	{
		"location" : 
		{
			"filename" : "test/testcases/necla/ex44.c",
			"loc" : 
			{
				"col" : 5,
				"line" : 33
			}
		},
		"type" : "BUF-01"
	},
	
	{
		"location" : 
		{
			"filename" : "test/testcases/necla/ex44.c",
			"loc" : 
			{
				"col" : 9,
				"line" : 34
			}
		},
		"type" : "BUF-01"
	},
	
	{
		"location" : 
		{
			"filename" : "test/testcases/necla/ex44.c",
			"loc" : 
			{
				"col" : 9,
				"line" : 37
			}
		},
		"type" : "BUF-01"
	},
	
	{
		"location" : 
		{
			"filename" : "test/testcases/necla/ex44.c",
			"loc" : 
			{
				"col" : 9,
				"line" : 38
			}
		},
		"type" : "BUF-01"
	},
	
	{
		"location" : 
		{
			"filename" : "test/testcases/necla/ex44.c",
			"loc" : 
			{
				"col" : 9,
				"line" : 39
			}
		},
		"type" : "BUF-01"
	},
	
	{
		"location" : 
		{
			"filename" : "test/testcases/necla/ex44.c",
			"loc" : 
			{
				"col" : 5,
				"line" : 41
			}
		},
		"type" : "BUF-01"
	},
	
	{
		"location" : 
		{
			"filename" : "test/testcases/necla/ex44.c",
			"loc" : 
			{
				"col" : 4,
				"line" : 55
			}
		},
		"type" : "BUF-01"
	},
	
	{
		"location" : 
		{
			"filename" : "test/testcases/necla/ex44.c",
			"loc" : 
			{
				"col" : 5,
				"line" : 32
			}
		},
		"type" : "REQ-01"
	},
	
	{
		"location" : 
		{
			"filename" : "test/testcases/necla/ex44.c",
			"loc" : 
			{
				"col" : 4,
				"line" : 55
			}
		},
		"type" : "REQ-01"
	}
]
